upload.py

```diff
@@ -33,55 +35,80 @@ def hrsize(size):
             return '%.0f%sB' % (size, suffix)
     return '%.0f%sB' % (size, 'Y')
 
+
+class HashingFile(file):
+    def __init__(self, file_path):
+        super(HashingFile, self).__init__(file_path, "w+b")
+        self.sha1 = hashlib.sha1()
+
+    def write(self, data):
+        self.sha1.update(data)
+        return file.write(self, data)
+
+    def get_hash(self):
+        return self.sha1.hexdigest()
+
+
+def getHashingFieldStorage(upload_dir):
+    class HashingFieldStorage(cgi.FieldStorage):
+        def make_file(self, binary=None):
+            self.open_file = HashingFile(os.path.join(upload_dir, self.filename))
```

Comment:
Not that it matters for this PR, but this should probably sanitize the filename.